sklearn_numba_dpex/common/kernels.py

```diff
@@ -223,43 +232,90 @@ def make_sum_reduction_2d_axis1_kernel(
+    summing. It must be a function that will be interpreted as a dpex.func and is
+    subject to the same rules. It is expected to take one scalar argument and returning
+    one scalar value. lambda functions are advised against since the cache will not
+    work with lambda functions. sklearn_numba_dpex.common._utils expose some
+    pre-defined `fused_elementwise_func`s.
 
     Notes
     -----
-    `work_group_size` is assumed to be a power of 2.
+    if `size1` is None then the kernel expects 1d tensor inputs, and `axis` parameter
+    is ignored.
+
+    If `size1` is not None then the expected shape of input tensors is
+    `(size0, size1)`, and the reduction operation is equivalent to
+    `input.sum(axis=axis)`.
+
+    If `size1` is None or if `axis` is `0`, then `work_group_size` is assumed to be a
+    power of 2, and the parameter `sub_group_size` is ignored.
+
+    If `size1` is not `None` and `axis` is `1`, then `work_group_size` is assumed to be
+    a multiple of `sub_group_size`, such that `work_group_size // sub_group_size` is
+    a power of 2.
+
+    Depending on the size of the sum axis, it might be worth tuning `work_group_size`
+    for better performance. When `axis` is `1`, tuning `sub_group_size` with respect
+    to the size of the other axis might also be beneficial.
+
+    The algorithmic approach that is chosen to implement the underlying kernels takes
+    inspiration from many sources that are available online, such as [1]_, [2]_, [3]_ .
+
+    .. [1] Timcheck, S. W. (2017). Efficient Implementation of Reductions on GPU
+    Architectures.
+
+    .. [2] https://dournac.org/info/gpu_sum_reduction
 
-    if `size1` is None then the kernel expects 1d tensor inputs. If `size1` is not None
-    then the expected shape of input tensors is `(size0, size1)`, and the reduction
-    operation is equivalent to input.sum(axis=1). Depending on the size of the sum
-    axis, it might be worth tuning `work_group_size` for better performance.
+    .. [3] https://shreeraman-ak.medium.com/parallel-reduction-with-cuda-d0ae10c1ae2c
     """
-    if size1 is not None:
-        n_rows = size0
+    if is_1d := not size1:
+        axis = 1
+        size1 = size0
+        size0 = 1
+
+    # NB: the shape of the work group is different in each of those two cases. Summing
+    # efficiently requires to adapt to very different IO patterns depending on the sum
+    # axis, which motivates the need for different kernels with different work group
+    # sizes for each cases. As a consequence, the shape of the intermediate results
+    # in the main driver loop, and the `global_size` (total number of work items fired
+    # per call) for each kernel call, are also differents, and are variabilized with
+    # the lambda functions `get_result_shape1 and `get_global_size`.
+    if axis == 1:
+        work_group_size, kernels = _prepare_sum_reduction_2d_axis1(
+            size0, work_group_size, fused_elementwise_func, dtype, device
+        )
```

Comment:
I find this unpacking unconventional and here my proposal to improve it (1/3).
```suggestion
        (
            work_group_size,
            kernels, 
            reduction_block_size,
        ) = _prepare_sum_reduction_2d_axis1(
```